manual/src/tutorials/memorymodel.etex

```diff
@@ -0,0 +1,740 @@
+frontier. The frontier models the synchronization information, which is merged
+with the frontiers of threads that operate on the location. In this way,
+non-atomic writes made by one thread can become known to another by
+communicating via an atomic location.
+
+For example,
+
+\begin{verbatim}
+(* Atomic locations *)
+A: 10, [a -> t1; b -> t5; c -> t7]
+B: 5,  [a -> t2; b -> t4; c -> t6]
+\end{verbatim}
+
+\noindent shows two atomic variables "A" and "B" with values "10" and "5",
+respectively, and frontiers of their own. We use upper-case variable names to
+indicate atomic locations.
+
+During atomic reads, the frontier of the location is merged into that of the
+domain performing the read. For example, suppose "d1" reads "B". The read
+returns "5", and "d1"'s frontier merged with "B"'s frontier, choosing the later
+timestamp for each location. The abstract machine state before the atomic read
+is:
+
+\begin{verbatim}
+(* Non-atomic locations *)
+a: [t1 -> 1; t2 -> 2]
+b: [t3 -> 3; t4 -> 4; t5 -> 5]
+c: [t6 -> 5; t7 -> 6; t9 -> 10; t8 -> 7]
+
+(* Domains *)
+d1: [a -> t1; b -> t3; c -> t6]
+d2: [a -> t1; b -> t4; c -> t9]
+
+(* Atomic locations *)
+A: 10, [a -> t1; b -> t5; c -> t7]
+B: 5,  [a -> t2; b -> t4; c -> t6]
+\end{verbatim}
+
+As a result of the atomic read, the abstract machine state is updated to:
+
+\begin{verbatim}
+(* Non-atomic locations *)
+a: [t1 -> 1; t2 -> 2]
+b: [t3 -> 3; t4 -> 4; t5 -> 5]
+c: [t6 -> 5; t7 -> 6; t9 -> 10; t8 -> 7]
+
+(* Domains *)
+d1: [a -> t2; b -> t4; c -> t6] (* frontier updated at a and b *)
+d2: [a -> t1; b -> t4; c -> t9]
+
+(* Atomic locations *)
+A: 10, [a -> t1; b -> t5; c -> t7]
+B: 5,  [a -> t2; b -> t4; c -> t6]
+\end{verbatim}
+
+During atomic writes, the value held by the atomic location is updated. The
+frontiers of both the reading domain and that of the location being read is
+updated to the merge of the two frontiers. For example, if "d2" writes "20" to
+"A" in the current machine state, the machine state is updated to:
+
+\begin{verbatim}
+(* Non-atomic locations *)
+a: [t1 -> 1; t2 -> 2]
+b: [t3 -> 3; t4 -> 4; t5 -> 5]
+c: [t6 -> 5; t7 -> 6; t9 -> 10; t8 -> 7]
+
+(* Domains *)
+d1: [a -> t2; b -> t4; c -> t6]
+d2: [a -> t1; b -> t5; c -> t9] (* frontier updated at b *)
+
+(* Atomic locations *)
+A: 20, [a -> t1; b -> t5; c -> t9] (* value updated. frontier updated at c. *)
+B: 5,  [a -> t2; b -> t4; c -> t6]
+\end{verbatim}
+
+\subsection{s:mm_semantics_reasoning}{Reasoning with the semantics}
+
+Let us revisit an example from earlier (section \ref{s:why_relaxed_memory}).
+
+\begin{caml_example*}{verbatim}
+let a = ref 0
+and b = ref 0
+
+let d1 () =
+  a := 1;
+  !b
+
+let d2 () =
+  b := 1;
+  !a
+
+let main () =
+  let h = Domain.spawn d2 in
+  let r1 = d1 () in
+  let r2 = Domain.join h in
+  assert (not (r1 = 0 && r2 = 0))
+\end{caml_example*}
+
+This program has a data race on "a" and "b", and hence, the program may observe
```

Comment:
```suggestion
This program has a data race on "a" and "b", and hence, the program may exhibit
```

Reply:
"Exhibit" is indeed the right term. Thanks.